bsip-0061.md

```diff
@@ -0,0 +1,169 @@
+    - Initiator (potentially *account_to_unlock*) signs the transaction using its previous authority preceding the locked state.
+    - Initiator (potentially *account_to_unlock*) pays fee for this operation.
+    - *account_to_unlock* should have sufficient balances to perform this operation.
+    - Check if *account_to_unlock* is locked in unlock_account_evaluator.
+    - Check if *account_to_unlock* is authorized to perform the operation (transaction signed with previous authority).
+    - Undo "lock"-operation. See Discussion - Unlocking account.
+
+# Specifications
+## Prevent circular dependencies
+```
+account_create_evaluator::do_evaluate( )
+{  ...
+   if( head_block_time() < HARDFORK_PREVENT_CYCLES_TIME )
+   {
+       graphene::chain::verify_cycled_authority( account );
+   }
+   ...
+}
+
+account_update_evaluator::do_evaluate( )
+{  ...
+   if( head_block_time() < HARDFORK_PREVENT_CYCLES_TIME )
+   {
+       graphene::chain::verify_cycled_authority( account );
+   }
+   ...
+}
+```
+```
+verify_cycled_authority( account_id )
+{
+   sign_state state( initial_parameters, verify_only=true );
+
+   GRAPHENE_ASSERT( state.check_authority(account_id), "Missing Authority" );
+}
+```
+```
+class sign_state
+{
+    sign_state( old_parameters, bool verify_only = false )
+
+    //
+    // New method added to parametrize class behaviour
+    //
+    bool is_key_available(const public_key_type &key)
+    {
+        if (verify_only)
+        {
+           return true
+        }
+
+        auto pk = available_keys.find(key);
+        return (pk != available_keys.end());
+    }
+
+    ...
+    bool verify_only
+}
+```
+
+# Discussion
+## Unlocking account
+Operation that prior to locked state will be fully re-done. There is an option - only owner or active authorities can be restored.
+
+# Examples
+## Unlocking account using previous authority. Fully re-done operation
+There is following use case:
+
+    1) Account "Alice" was created
+    2) Account "Alice" was updated with new authorities and options (opts 1) - here: prior state
+    3) Account "Alice" was updated with new options (opts 2)
+    4) Account "Alice" was updated with new options (opts 3)
+    5) Account "Alice" was updated with new authorities - here: locked state
+
+To unlock "Alice" we need to reapply update operation at step 2. But latest account options (opts 3) at step 4 will be lost.
+
+## Unable to unlock account
+Use case:
+
+    1) Account "Alice" delegates its authority to "Bob" - here: prior state
+    2) Account "Charlie" delegates its authority to "Alice"
+    3) Account "Alice" delegates its authority to "Charlie" (transaction was signed with Bob's private key) - here: locked state
+    4) Account "Bob" delegates its authority to "Charlie" - here: locked state
+
+To unlock "Alice" we need to reapply update operation at step 1. But "Bob" also locked, we have double lock here.
```

Comment:
The algorithm I outlined above would detect the locks on Alice and Bob. It would also resurrect both, although it would be sufficient to resurrect Alice. I think that's acceptable. Opinions?

Reply:
Here I meant that account recovery in one step is not always possible. See diagram below.

![Unlock accounts](https://user-images.githubusercontent.com/42674402/55235482-b5058a80-523e-11e9-82fd-90f34d0d8eaf.png)

So, automatic recovery requires repeating unlock procedure to a normal state. In manual mode, another operation may also be required before full unlocking.

Reply:
Right, thanks.

Reply:
In the first the first transition (3), Alice changes her authority from Bob to Charlie, which creates a cycle. I thought we had guards on that to a certain depth, no? Same for Bob: do we guard against him delegating into an cycle. 

Why does the undo happen in the forward order 3, then 4? I thought it would undo the most recent and repeat until the cycle is eliminated.

Reply:
@ryanRfox :
1. we need to fix historical locked accounts, the logic is described above by @OpenLedgerApp ;
2. after that, in the long run, we avoid new account locking with the logic described by you.

Reply:
Thanks, abitmore!
This is exactly what we meant.